segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java

```diff
@@ -173,75 +200,68 @@ public int getId() {
      * @param operations The initial set of operations to process (in order). Multiple operations may be processed eventually
      *                   depending on how the operationQueue changes while this is processing.
      */
+    @SneakyThrows
     private void processOperations(Queue<CompletableOperation> operations) {
         log.debug("{}: processOperations (OperationCount = {}).", this.traceObjectId, operations.size());
 
-        // Create a new State and Builder (we need this either initially or after recovery from an error).
-        final QueueProcessingState state = new QueueProcessingState(this.metadataUpdater, this.stateUpdater, this.checkpointPolicy, this.traceObjectId);
-        final DataFrameBuilder<Operation> dataFrameBuilder = new DataFrameBuilder<>(this.durableDataLog, state::commit, state::fail);
+        // Process the operations in the queue. This loop will ensure we do continuous processing in case new items
+        // arrived while we were busy handling the current items.
+        while (!operations.isEmpty()) {
+            try {
+                // Get the DataFrameBuilder, and recover it if necessary.
+                val builder = getDataFrameBuilder(true);
 
-        try {
-            // Process the operations in the queue. This loop will ensure we continue processing after a recoverable failure,
-            // as well as after we processed the entire collection, but found more items in need of processing.
-            while (!operations.isEmpty()) {
                 // Process the current set of operations.
-                processOperations(operations, state, dataFrameBuilder);
+                while (!operations.isEmpty()) {
+                    CompletableOperation o = operations.poll();
+                    if (processOperation(o, builder)) {
```

Comment:
The operation is already failed with the proper exception. It is up to the user (I.e.`AppendProcessor`, `PravegaRequestProcessor` to log the exception that it failed with). There is no point in double-logging it here.